File: 3_LongestSubstringWithoutRepeatingCharacters.js
/**
 * @param {string} s
 * @return {number}
 */
const lengthOfLongestSubstring = function (s) {
    //此用來存放每一個字元最後出現的索引
    const map = {};
    //此為當前不重複字串的開頭索引
    let left = 0;

    //將傳進字串切割變為陣列 like['a','b','c']
    return s.split('').reduce((max, value, index) => {
        /*比較每個字元出現的索引是否已經有值且大於等於 開頭索引  map[index]  不為undefined時 代表重複  故須以當前字元索引+1 
         *當作此切割字串的開頭索引
         */
        left = map[value] >= left ? map[value] + 1 : left;

        //紀錄每個字元當前最新的索引
        map[value] = index
        //比較max  與字串長度並回傳
        return Math.max(max, index - left + 1);
    }, 0)
};

File: 3.LongestSubstringWithoutRepeatingCharacters/3_LongestSubstringWithoutRepeatingCharacters.py
# 解法1
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # 傳入空字串的情形
        if len(s) == 0:
            return 0

        # 存放每一個字元最後出現的索引
        dic = {}
        # 不重複字串的開頭索引
        left = 0
        # 最大長度至少為1
        maxLength = 1

        for index in range(len(s)):
            # 若當前字元最後出現索引 大於等於不重複開頭  更新開頭索引
            # 當前字元不在dic內  不執行if判斷
            if s[index] in dic:
                if dic[s[index]] >= left:
                    left = dic[s[index]] + 1

            # 紀錄每個字元最新索引
            dic[s[index]] = index
            # 更新最大長度
            maxLength = max(maxLength, index - left + 1)

        return maxLength

test = Solution()
print(test.lengthOfLongestSubstring("abcabcbb"))
